lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -405,18 +341,22 @@ handleHeadersCommunication
 handleHeadersCommunication logic conv = do
     whenJustM (recvLimited conv) $ \mgh@(MsgGetHeaders {..}) -> do
         logDebug $ sformat ("Got request on handleGetHeaders: "%build) mgh
-        -- FIXME recoveryInProgress is a logic layer notion; get it from the
-        -- logic layer itself. Better yet: don't use it at all. Diffusion layer
-        -- is entirely capable of serving blocks even if the logic layer is in
-        -- recovery mode.
+        -- FIXME 
+        -- Diffusion layer is entirely capable of serving blocks even if the
```

Comment:
This reasoning lacks proofs. It's not obvious that the data we would serve while being in recovery mode is sane and would not break the protocol. I can describe scenario in which it would. In general, it wasn't explicitly discussed, so it is banned by default.

Reply:
> It's not obvious that the data we would serve while being in recovery mode is sane and would not break the protocol.  I can describe scenario in which it would.

Please do.

Reply:
Ok, I've spent some time trying to figure it out, and i don't have an exact answer. I was thinking of the following. Big amount of nodes in recovery can temporary switch to non-optimal branch which was abandoned because it's not the longest one (but the longest one relatively to the current synced tip, which is old). If they serve `tip` to newcomers, newcomers will sync to some stuff as well. Malicious agent can promise some tip X which is the most difficult one, but provide [curTip...b] to clients which will not lead to X itself. Thus, a lot of nodes can be keeped in the state, where they can't be synced with the main chain, but think they will. If it happens for `2k` slots and for big part of the stake, protocol will fail. I cannot really assume anything here, I guess, because my knowledge of protocol mechanics (of proofs, common string etc) is very poor.

Reply:
I think such an attack is already possible in `master`/`develop`. An attacker can make up a header which is significantly more difficult that the real tip, present it to everyone, and they'll all go into useless recovery mode, and nobody will be able to receive blocks (nor will any blocks be minted). That's even worse, no?

Reply:
It's a complex question and should be discussed in CSL-1514. I can give my thoughts on it, but it's not related for now.